tools/openslide-quickhash1sum.1.in
.\"
.\" OpenSlide, a library for reading whole slide image files
.\"
.\" Copyright (c) 2007-2012 Carnegie Mellon University
.\" Copyright (c) 2023      Benjamin Gilbert
.\" All rights reserved.
.\"
.\" OpenSlide is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU Lesser General Public License as
.\" published by the Free Software Foundation, version 2.1.
.\"
.\" OpenSlide is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
.\" GNU Lesser General Public License for more details.
.\"
.\" You should have received a copy of the GNU Lesser General Public
.\" License along with OpenSlide. If not, see
.\" <http://www.gnu.org/licenses/>.
.\"


.\" See man-pages(7) for formatting conventions.


.TH OPENSLIDE\-QUICKHASH1SUM 1 2023-06-30 "OpenSlide @SUFFIXED_VERSION@" "User Commands"

.mso www.tmac

.SH NAME
openslide\-quickhash1sum \- Print OpenSlide quickhash-1 checksums

.SH SYNOPSIS
.BR "openslide\-quickhash1sum " [ \-\-help "] [" \-\-version ]
.IR slide ...

.SH DESCRIPTION
Print OpenSlide
.I quickhash-1
checksums for one or more virtual slide files, in a format similar to
.BR sha256sum (1).
.PP
.I quickhash-1
is a non-cryptographic, 256-bit hash of a subset of a slide's data.
It uniquely identifies a particular virtual slide,
but cannot be used to detect corruption or modification of the slide file.

.SH OPTIONS
.TP
.B \-\-help
Display usage summary.

.TP
.B \-\-version
Display version and copyright information.

.SH EXIT STATUS
.B openslide\-quickhash1sum
returns 0 on success, 1 if a
.I quickhash-1
could not be calculated, or 2 if the arguments are invalid.

.SH NOTES
.B openslide\-quickhash1sum
is a legacy command and no new features will be added.
Its replacement is the
.B slide quickhash1
subcommand of
.BR slidetool (1).

.SH BUGS
.I quickhash-1
is not defined for all slide files supported by OpenSlide.
If a slide does not have a
.IR quickhash-1 ,
"No quickhash\-1 available" will be printed.

.SH COPYRIGHT
Copyright \(co 2007-2026 Carnegie Mellon University and others
.PP
OpenSlide is free software: you can redistribute it and/or modify it
under the terms of the
.URL http://gnu.org/licenses/lgpl-2.1.html "GNU Lesser General Public License, version 2.1" .
.PP
OpenSlide comes with NO WARRANTY, to the extent permitted by law.
See the GNU Lesser General Public License for more details.

.SH SEE ALSO
.BR sha256sum (1),
.BR slidetool (1)


CHANGELOG.md
# Notable Changes in OpenSlide


## Version 4.0.0, 2023-10-11

### Breaking changes

* Update soname to `libopenslide.so.1`
* Remove all deprecated functions
* Clear `openslide_read_associated_image()` output buffer on error
* Remove undocumented NULL `dest` support in `openslide_read_associated_image()`
* Change GLib log domain to `OpenSlide`
* Convert build system to Meson (thanks, Jan Harkes)

### New features

* New format: DICOM WSI (thanks, John Cupitt and Jim O'Donnell)
* Add APIs to read ICC color profiles (thanks, John)
* Add APIs to configure tile cache size and share caches between slides
* Add properties for associated image metadata
* generic-tiff: Set MPP properties if available
* philips: Set objective power property if available
* Add `slidetool` command-line tool which supports all OpenSlide features
* Combine all command-line tools into the same binary
* Add ICC profile to PNG images written by command-line tools
* Add self-test that doesn't require sample data (run with `meson test`)

### Changes

* Require libtiff ≥ 4, OpenJPEG ≥ 2.1, GLib ≥ 2.56, SQLite ≥ 3.14
* Require libdicom ≥ 1.0, with build-time fallback for now
* Remove support for including `openslide.h` in Visual Studio \< 2013
* Fail `openslide_open()` with a broken pixman 0.38.x
* Avoid extra buffer copy in `openslide_read_region()`
* Replace `goto`-based cleanup with `g_autoptr`
* Stop using deprecated GLib slice allocator
* Use internal wrappers for file I/O
* Documentation improvements

### Bug fixes

* Use UTF-8 filenames on Windows
* Improve `openslide-write-png` performance for very large regions
* Fix assertions on JPEG decode errors when compiled with Clang
* Portability fixes (thanks, Billy Robert O'Neal III and Kleis Auke Wolthuizen)
* aperio: Set objective power property even if floating point
* hamamatsu: Fix `Restart marker not found` on VMS slides with multiple Z-layers
* hamamatsu: Fix integer overflow in VMS parsing (thanks, Adam Goode)
* mirax: Fix `Expected 1 value` error
* philips: Don't cache missing tiles
* sakura: Fix memory leak reading missing tile


## Version 3.4.1, 2015-04-20

* New formats: Philips TIFF, Ventana TIFF
* Support OpenJPEG 2.1.0
* Improve performance of JPEG and JP2K decoding
* Add `openslide.region[i].*` properties
* Improve MATLAB compatibility
* Enable function deprecation warnings with MSVC
* Many portability fixes
* aperio: Detect OpenJPEG chroma subsampling breakage during `open`
* aperio: Fill in missing tiles with downsampled data
* aperio: Report MPP for slides scanned in locales with decimal comma
* hamamatsu: Support NDPI files \> 4 GB
* hamamatsu: Properly detect NDPI slides produced by NDP.toolkit
* hamamatsu: Support VMS/VMU slides without a `NoLayers` key
* hamamatsu: Report MPP for VMS/VMU
* leica: Support slides with `2010/03/10` XML namespace
* leica: Base64-decode `leica.barcode` property in `2010/10/01` namespace
* sakura: Support slides with multiple focal planes
* sakura: Support slides without tile table
* ventana: Support slides with multiple focal planes
* ventana: Improve positioning of AOIs within level
* ventana: Fix failure to recognize macro image on some slides


## Version 3.4.0, 2014-01-25

* Major internal restructuring
* New formats: Hamamatsu NDPI, Sakura SVSLIDE, Ventana BIF (preliminary)
* Add `openslide_detect_vendor()`
* Deprecate `openslide_can_open()` (not very useful and often misused)
* Document performance considerations for `openslide_open()`
* Add properties giving the bounds of the non-empty region of the slide
* leica: Support multiple main images if their levels are coplanar
* leica: Use slide size as level size
* mirax: Support PNG- and BMP-formatted slides
* mirax: Fix `Expected first 0 value` error opening some slides
* mirax: Fix incorrect tile placement on some slides without overlaps
* mirax: Never synthesize downsampled levels
* Add `OPENSLIDE_DEBUG` environment variable (`OPENSLIDE_DEBUG=?` for help)
* Fix some crashes in error paths
* Add tests for many error paths


## Version 3.3.3, 2013-04-13

* Fix inclusion of `openslide.h` with MSVC
* Properly handle Aperio JP2K slides with zero-length tiles
* Support Hamamatsu slides with blank `MacroImage` key


## Version 3.3.2, 2012-12-01

* Fix seams in MIRAX 2.2 slides (thanks, Agelos Pappas)
* Fix associated image naming in single-level Aperio slides
* Stop decoding MIRAX tiles outside requested region
* Stop decoding unneeded tiles during tile-aligned accesses
* Increase Hamamatsu VMU tile size to reduce rendering overhead
* Document performance considerations for `openslide_can_open()`


## Version 3.3.1, 2012-10-14

* Parallelize concurrent `openslide_read_region` calls on an `openslide_t`
* Eliminate background scanning of tile headers in MIRAX
* Scan many fewer tiles during first accesses to Hamamatsu VMS
* Ignore Leica Z-planes other than 0
* Add experimental tile-size properties
* Document API thread safety


## Version 3.3.0, 2012-09-08

* Support Leica SCN format (requires libtiff ≥ 4) (thanks, Agelos Pappas)
* Allow opening MIRAX 2.2 slides (though there are seams, bug #92)
* Add standardized microns-per-pixel and objective-power properties
* Add `macro` associated image in Trestle
* Rename `layer` to `level` throughout the API (deprecate `layer` functions;
  remove `layer` properties)
* Report parse errors in `openslide_open()` by returning an `openslide_t` in
  error state
* Deprecate `openslide_get_comment()`
* Add `openslide_get_version()`
* Improve command-line tools; add manpages
* Support building with MinGW-w64; drop CMake, MSVC, mingw32
* Add tests for many error paths


## Version 3.2.6, 2012-02-23

* Support downsampled MIRAX files
* Improve performance on MIRAX slides without tile overlaps
* Fix `openslide_read_region` for large dimensions on layer \> 0
  (3.2.5 regression)
* Correct subpixel error in MIRAX tile placement
* Fix unlikely use-after-free with Hamamatsu VMU


## Version 3.2.5, 2011-12-16


tools/openslide-quickhash1sum.c
/*
 *  OpenSlide, a library for reading whole slide image files
 *
 *  Copyright (c) 2010-2012 Carnegie Mellon University
 *  All rights reserved.
 *
 *  OpenSlide is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as
 *  published by the Free Software Foundation, version 2.1.
 *
 *  OpenSlide is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with OpenSlide. If not, see
 *  <http://www.gnu.org/licenses/>.
 *
 */

#include <stdbool.h>
#include <stdio.h>
#include <glib.h>
#include "openslide.h"
#include "openslide-common.h"

static bool process(const char *file) {
  g_autoptr(openslide_t) osr = openslide_open(file);
  if (common_warn_on_error(osr, "%s", file)) {
    return false;
  }

  const char *hash =
    openslide_get_property_value(osr, OPENSLIDE_PROPERTY_NAME_QUICKHASH1);
  if (hash == NULL) {
    common_warn("%s: No quickhash-1 available", file);
    return false;
  }

  printf("%s  %s\n", hash, file);
  return true;
}


static const struct common_usage_info usage_info = {
  "FILE...",
  "Print OpenSlide quickhash-1 (256-bit) checksums.",
};

int main (int argc, char **argv) {
  common_parse_commandline(&usage_info, &argc, &argv);
  if (argc < 2) {
    common_usage(&usage_info);
  }

  int ret = 0;
  for (int i = 1; i < argc; i++) {
    if (!process(argv[i])) {
      ret = 1;
    }
  }

  return ret;
}
